### DISCUSSION TITLE
How to represent received sysex

### DISCUSSION THREAD

**Discussion by jojogib** [RomanKubiak/ctrlr#544] [Q&A] [answered]
Hey eveyone!
I want to represent the value of the incoming sysex on this modulator.
Like is if the incoming message matches with this modulator update value.
I need this for every modulator I will do later on. 
Whats the best/simplest methode?
Thank you 
[howtoreceive.zip](https://github.com/RomanKubiak/ctrlr/files/10989154/howtoreceive.zip)

**Comment by thethirdapple** [comment]
Greetings,

Which version of Ctrlr are you using?
The attached file returns a few xml related error and is not opening in the various versions of Ctrlr which I have tried...

**Comment by damiensellier** [comment, marked as answer]
Salut Jojo, your panel crashes CTRLR.
I didn't understand exactly what you need.
A sysex message is received by ctrlr, you want to update a specific modulator value from the sysex message right?
It starts with a LUA script assigned the global panel properties "**when a panel receives a midi message**":

CLICK THE CANVAS and OPEN THE PANEL MODE cmd+E, it's in the middle:
![CLICK THE CANVAS and OPEN THE PANEL MODE cmd+E](https://user-images.githubusercontent.com/55164385/219828935-ffe19cf6-3949-45d0-b4a2-7f8052770e55.png)

This script has to filter to listen only sysex messages, decode it (to get the values of the different parameters), then update the values of the corresponding panel modulators. 

to filter sysex messages on input it works with : 
`if midi:getType() == 5 then -- it is a SYSEX message`

The rest is almost the same as in the following solution by John @dnaldoog but this is for for cc values. You'll need to locate the parameter position and its value in the message though:
https://github.com/RomanKubiak/ctrlr/discussions/541#discussioncomment-5112997

**Comment by jojogib** [comment]
Thank you both i got it to work. I have one futher question. I have a very large sysex file > 20kb here is the beginning:

`F0 00 20 32 7F 15 20 00 00 24 72 65 76 20 52 31 20 3B 20 46 69 72 6D 77 61 72 65 20 31 2E 31 30 3B 20 42 43 20 4D 61 6E 61 67 65 72 20 34 2E 31 2E 32 F7 F0 00 20 32 7F 15 20 00 01 24 67 6C 6F 62 61 6C F7 F0 00 20 32 7F 15 20 00 02 20 20 2E 6D 69 64 69 6D 6F 64 65 20 55 2D 33 F7 F0 00 20 32 7F 15 20 00 03 20 20 2E 73 74 61 72 74 75 70 20 6C 61 73 74 F7 F0 00 20 32 7F 15 20 00 04 20 20 2E 66 6F 6F 74 73 77 20 61 75 74 6F F7`

I wanna do a button to send the whole message at once but I need a buffer like a delay between every F7 F0 . Anyone knows how to do this in lua? Thank you

**Reply by damiensellier** [reply]
F0 is the beginning of a sysex message F7 is the end of transmission. It means that in your 20kb file, there's a packet of several sysex messages. to introduce a delay between those shorter messages, you would first need a timer loop to send the different 51 bytes long messages. 
1st pass icount= 0 : send a 51bytes long message from byte (51x0) -- first byte position is 0 
2nd pass icount=1 : send a 51bytes long message from byte (51x1) -- 
3rd pass icount=2 : send a 51bytes long message from byte (51x2) -- 
...
you can see the pattern related to the icount iteration starting at icount = 0

Also the class getRange is perfect to split the messages : getRange(start_position_starts_at_0, number_of_bytes)
https://ctrlr.org/forums/topic/getrange-how-does-it-work/
https://ctrlr.org/forums/topic/getsize-only-looking-at-last-byte-number/#post-72758

Your first range is : midiMessage:getData():getRange(0,51)
Your 2nd range is : midiMessage:getData():getRange(51,51)
...

so it's all about the loop counter, incrementing its value+1 every pass, finding the right pattern for the byte position (start, length) and the delay between the timer loop passes.

the rest is in the following thread, I needed that as well to send yamaha sysex program dumps : 
https://github.com/RomanKubiak/ctrlr/discussions/285#discussioncomment-708015
https://github.com/RomanKubiak/ctrlr/discussions/442#discussioncomment-2307359

**Reply by jojogib** [reply]
Hey, because its every time the same sysex  I "solved" it in google sheets so i get something like this:

```
send = function(mod, value, source)
    local mStr1 = string.format("F0 00 20 32 7F 15 20 00 00 24 72 65 76 20 52 31 20 3B 20 46 69 72 6D 77 61 72 65 20 31 2E 31 30 3B 20 42 43 20 4D 61 6E 61 67 65 72 20 34 2E 31 2E 32 F7")
    panel:sendMidiMessageNow(CtrlrMidiMessage(mStr1))
os.execute(sleep(2))
    local mStr2 = string.format("F0 00 20 32 7F 15 20 00 01 24 67 6C 6F 62 61 6C F7")
    panel:sendMidiMessageNow(CtrlrMidiMessage(mStr2))

os.execute(sleep(2))
    local mStr3 = string.format("F0 00 20 32 7F 15 20 00 02 20 20 2E 6D 69 64 69 6D 6F 64 65 20 55 2D 33 F7")
    panel:sendMidiMessageNow(CtrlrMidiMessage(mStr3))

os.execute(sleep(2))
    local mStr4 = string.format("F0 00 20 32 7F 15 20 00 03 20 20 2E 73 74 61 72 74 75 70 20 6C 61 73 74 F7")
    panel:sendMidiMessageNow(CtrlrMidiMessage(mStr4))
end
```
but now the os.execute(sleep(2)) will not execute after every message it seems like it will first add all the sleep seconds and after it will send it all together.  Any solution for this? Thank you

**Reply by damiensellier** [reply]
os.execute() is not really recommanded it locks all the system for the amount of time.
You set it to (2) which is 2ms, this is very short IMHO unless your receiving gear is able to handle it.
Try os.execute (500) for 500ms to see if the messages are properly delayed with this kind of interval inbetween.

**Reply by jojogib** [reply]
yeah i know i tried diffrent values but always with the same result

**Reply by thethirdapple** [reply]
Greetings,

@jojogib, would you mind sharing your solution to “parsing” of a long sysex string? 

The Korg Karma panel i am working thru would be greatly improved with something similar. Its easy enough to switch banks and presets with simple external sysex message, but i noticed that when using the on board buttons to switch banks, the keyboard generates a super long string containing parameter settings relating to the preset…

peace

**Reply by dnaldoog** [reply]
> Greetings,
> 
> @jojogib, would you mind sharing your solution to “parsing” of a long sysex string?
> 
> The Korg Karma panel i am working thru would be greatly improved with something similar. Its easy enough to switch banks and presets with simple external sysex message, but i noticed that when using the on board buttons to switch banks, the keyboard generates a super long string containing parameter settings relating to the preset…
> 
> peace

Hi @thethirdapple 

If you are looking to parse incoming Korg sysex, which is a nightmare, I have worked out lua algorithms for rejigging the data i.e Converting "MIDI 7-Bit Data" to "8-Bit Data". I am guessing the Karma uses the same system as other Korg machines.

**Reply by thethirdapple** [reply]
That sounds super helpful, thank you. I will take another look at the Karma strings it generates and post one in a new thread(dont want to hijack jojgib’s focused questions), its something I didn't want to get stuck on as it is not crucial for my “personal” needs, but would be nifty if the  panel gets to a state of public consumption…

**Reply by jojogib** [reply]
> Greetings,
> 
> @jojogib, would you mind sharing your solution to “parsing” of a long sysex string?
> 
> The Korg Karma panel i am working thru would be greatly improved with something similar. Its easy enough to switch banks and presets with simple external sysex message, but i noticed that when using the on board buttons to switch banks, the keyboard generates a super long string containing parameter settings relating to the preset…
> 
> peace
Sorry for my english I dont know the exact terms for google sheets in english.
. I paste the full sysex in goole sheets in one "chamber". Then I used split by F7, so every sysex starts by F0 in a single row.  Now F7 is missing because of the spilt so I made one row of F7. For every paramterer I like to change I used single row. Sorry its a little hard to explain. I will post the sheet here.
[Unbenannte Tabelle.xlsx](https://github.com/RomanKubiak/ctrlr/files/11049605/Unbenannte.Tabelle.xlsx)

**Comment by dnaldoog** [comment]
It looks like your sysex in the file has different sizes. You need something [like](https://sourceforge.net/projects/ctrlr-utilities/files/send%20sysex%20from%20file/):

`buf` is the binary contents of a sysex file:

<pre>
sendBufferContents = function()
panel:setProperty("panelMidiGlobalDelay", 300, true)
    local start, eox = 0, 0 -- initialise
    for i = 0,buf:getSize() - 1 do
        if buf:getByte(i) == 240 then -- F0 start of message
            start = i
        elseif buf:getByte(i) == 247 then -- EOX F7 end of message
            eox = i
            panel:sendMidiMessageNow(CtrlrMidiMessage(buf:getRange(start, eox - start):toHexString(1)))
        end
    end
end
</pre>

You don't need a timer to delay between messages - you can use:

`panel:setProperty("panelMidiGlobalDelay", value, true)` where `value` = delay in milliseconds.

**Reply by damiensellier** [reply]
You have good eyes my friend! I assumed the messages were all the same size, I'm impressed you had this idea :) 
I don't like the midi global delay because if you update a cc value in real time even with 10ms from 0 to 127 at the end you have more than a second of delay. for sysex dumps like this it's ok but after the dump it has to be reset to 0 then to avoid future lag.

**Reply by dnaldoog** [reply]
> You have good eyes my friend! I assumed the messages were all the same size, I'm impressed you had this idea :) I don't like the midi global delay because if you update a cc value in real time even with 10ms from 0 to 127 at the end you have more than a second of delay. for sysex dumps like this it's ok but after the dump it has to be reset to 0 then to avoid future lag.

I thought they were all 51 bytes in length when I first saw it too!

I see what you mean about the global delay thing. It's a new discovery for me, but [here is some altered code]( https://sourceforge.net/projects/ctrlr-utilities/files/send%20sysex%20from%20file/) using a timer &#8212; code adapted from a Roland panel of mine that also has variable sized messages inside a dump.

<pre>
sendBufferContents = function(comp, mouse)
    if fileContentsBuffer:getSize() == 0 then return end
    local t = {}
        local j = 0
        for i = 0, fileContentsBuffer:getSize() - 1 do
            if fileContentsBuffer:getByte(i) == 247 then
                table.insert(t, fileContentsBuffer:getRange(j, (i - j) + 1):toHexString(1))
                j = i + 1
            end
        end

    global_counter = 0
    global_tab = t
    global_delay = panel:getModulatorByName("setTransmitDelay"):getModulatorValue() -- milliseconds

    timer:setCallback(1, timerCallback)
    timer:startTimer(1, global_delay) 
end

timerCallback = function(timerId)
    if timerId == 1 then
        global_counter = global_counter + 1 -- increment global counter
        panel:sendMidiMessageNow(CtrlrMidiMessage(global_tab[global_counter]))
        if global_counter == #global_tab then
            timer:stopTimer(timerId)
        end
    end
end
</pre>

**Reply by jojogib** [reply]
Ok thank you I just checkd the panel and its almost what i like to do. But id like to keep it simpler I only need one button and want to predefine delay and the whole sysex string direct in lua. Is it possible to modify this section:
  ```
 local buf = MemoryBlock()
    fileRead:loadFileAsData(buf)
    fileContentsBuffer = buf
    panel:getLabel("loadL"):setText(fileContentsBuffer:getSize() .. " bytes")
end --function
```
like buf = F0 00 20 32 7F 15 20 00 00 24 72 65 76 20 52 31 20 3B 20 46 69 72 6D 77 61 72 65 20 31 2E 31 30 3B 20 42 43 20 4D 61 6E 61 67 65 72 20 34 2E 31 2E 32 F7 F0 00 20 32 7F 15 20 00 01 24 67 6C 6F 62 61 6C F7 F0 00 20 32 7F 15 20 00 02 20 20 2E 6D 69 64 69 6D 6F 64 65 20 55 2D 33 F7 F0 00 20 32 7F 15 20 00 03 20 20 2E 73 74 61 72 74 75 70 20 6C 61 73 74 F7 F0 00 20 32 7F 15 20 00 04 20 20 2E 66 6F 6F 74 73 77 20 61 75 74 6F F7 ? thank you

**Reply by dnaldoog** [reply]
> Ok thank you I just checkd the panel and its almost what i like to do. But id like to keep it simpler I only need one button and want to predefine delay and the whole sysex string direct in lua. Is it possible to modify this section:
> 
> ```
> local buf = MemoryBlock()
>   fileRead:loadFileAsData(buf)
>   fileContentsBuffer = buf
>   panel:getLabel("loadL"):setText(fileContentsBuffer:getSize() .. " bytes")
> end --function
> ```
> 
> like buf = F0 00 20 32 7F 15 20 00 00 24 72 65 76 20 52 31 20 3B 20 46 69 72 6D 77 61 72 65 20 31 2E 31 30 3B 20 42 43 20 4D 61 6E 61 67 65 72 20 34 2E 31 2E 32 F7 F0 00 20 32 7F 15 20 00 01 24 67 6C 6F 62 61 6C F7 F0 00 20 32 7F 15 20 00 02 20 20 2E 6D 69 64 69 6D 6F 64 65 20 55 2D 33 F7 F0 00 20 32 7F 15 20 00 03 20 20 2E 73 74 61 72 74 75 70 20 6C 61 73 74 F7 F0 00 20 32 7F 15 20 00 04 20 20 2E 66 6F 6F 74 73 77 20 61 75 74 6F F7 ? thank you

So the code 
<pre>
local buf = MemoryBlock()
  fileRead:loadFileAsData(buf)
  fileContentsBuffer = buf
</pre>

is for loading a binary sysex file off the computer.

If you just want that static string every time then create a function that runs when you click on a button that sends that code:



> Ok thank you I just checkd the panel and its almost what i like to do. But id like to keep it simpler I only need one button and want to predefine delay and the whole sysex string direct in lua. Is it possible to modify this section:
> 
> ```
> local buf = MemoryBlock()
>   fileRead:loadFileAsData(buf)
>   fileContentsBuffer = buf
>   panel:getLabel("loadL"):setText(fileContentsBuffer:getSize() .. " bytes")
> end --function
> ```
> 
> like buf = F0 00 20 32 7F 15 20 00 00 24 72 65 76 20 52 31 20 3B 20 46 69 72 6D 77 61 72 65 20 31 2E 31 30 3B 20 42 43 20 4D 61 6E 61 67 65 72 20 34 2E 31 2E 32 F7 F0 00 20 32 7F 15 20 00 01 24 67 6C 6F 62 61 6C F7 F0 00 20 32 7F 15 20 00 02 20 20 2E 6D 69 64 69 6D 6F 64 65 20 55 2D 33 F7 F0 00 20 32 7F 15 20 00 03 20 20 2E 73 74 61 72 74 75 70 20 6C 61 73 74 F7 F0 00 20 32 7F 15 20 00 04 20 20 2E 66 6F 6F 74 73 77 20 61 75 74 6F F7 ? thank you

Yes, like [this](https://github.com/RomanKubiak/ctrlr/files/11050037/send.sysex.from.file_2_1.zip): Just change the string `str` as required:


<pre>
sendBufferContents = function(comp, mouse)
local buf=MemoryBlock()
buf:loadFromHexString(str)
    if buf:getSize() == 0 then return end
    local t = {}
        local j = 0
        for i = 0, buf:getSize() - 1 do
            if buf:getByte(i) == 247 then
                table.insert(t, buf:getRange(j, (i - j) + 1):toHexString(1))
                j = i + 1
            end
        end

    global_counter = 0
    global_tab = t
    global_delay = panel:getModulatorByName("setTransmitDelay"):getModulatorValue() -- milliseconds
	panel:getLabel("loadL"):setText(buf:getSize().." bytes")
    timer:setCallback(1, timerCallback)
    timer:startTimer(1, global_delay) 
end

timerCallback = function(timerId)

    if timerId == 1 then
    	global_counter = global_counter + 1 -- increment global counter
        panel:sendMidiMessageNow(CtrlrMidiMessage(global_tab[global_counter]))
        if global_counter == #global_tab then
            timer:stopTimer(timerId)
        end
    end
end

str="F0 00 20 32 7F 15 20 00 00 24 72 65 76 20 52 31 20 3B 20 46 69 72 6D 77 61 72 65 20 31 2E 31 30 3B 20 42 43 20 4D 61 6E 61 67 65 72 20 34 2E 31 2E 32 F7 F0 00 20 32 7F 15 20 00 01 24 67 6C 6F 62 61 6C F7 F0 00 20 32 7F 15 20 00 02 20 20 2E 6D 69 64 69 6D 6F 64 65 20 55 2D 33 F7 F0 00 20 32 7F 15 20 00 03 20 20 2E 73 74 61 72 74 75 70 20 6C 61 73 74 F7 F0 00 20 32 7F 15 20 00 04 20 20 2E 66 6F 6F 74 73 77 20 61 75 74 6F F7"
</pre>

**Reply by jojogib** [reply]
Many thanks. Thats what i was searching for. Thank you
